File: arrays/contiguous_array.rb
# @param {Integer[]} nums
# @return {Integer}
def find_max_length(nums)
    sum = 0
    max_l = 0
    mapper = {}

    nums.each_with_index do |num, i|
        sum += num == 0 ? -1 : 1
        if sum == 0
            max_l = i + 1
        elsif mapper.key?(sum)
            max_l = [max_l, i - mapper[sum]].max
        else
            mapper[sum] = i
        end

    end
    
    max_l
end

File: arrays/contiguous_array.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        total_sum = 0
        max_length = 0
        sum_map = {}
        
        for i, num in enumerate(nums):
            total_sum += -1 if num == 0 else 1
            
            if total_sum == 0:
                max_length = i + 1
            elif total_sum in sum_map:
                max_length = max(max_length, i - sum_map[total_sum])
            else:
                sum_map[total_sum] = i
        
        return max_length